File: src-tauri/src/date_keys.rs
use anyhow::{Context, Result};
use chrono::{DateTime, Days, NaiveDate, Utc};
use chrono_tz::Tz;

pub fn parse_timestamp(value: &str) -> Option<DateTime<Utc>> {
    DateTime::parse_from_rfc3339(value)
        .ok()
        .map(|timestamp| timestamp.with_timezone(&Utc))
}

pub fn date_key_for(date: DateTime<Utc>, time_zone: &str) -> Result<String> {
    let time_zone: Tz = time_zone
        .parse()
        .with_context(|| format!("unsupported time zone: {time_zone}"))?;
    Ok(date
        .with_timezone(&time_zone)
        .format("%Y-%m-%d")
        .to_string())
}

pub fn add_days_to_date_key(date_key: &str, delta_days: i64) -> Result<String> {
    let date = NaiveDate::parse_from_str(date_key, "%Y-%m-%d")
        .with_context(|| format!("unsupported date key: {date_key}"))?;

    let shifted = if delta_days >= 0 {
        date.checked_add_days(Days::new(delta_days as u64))
            .context("failed to add days to date key")?
    } else {
        date.checked_sub_days(Days::new(delta_days.unsigned_abs()))
            .context("failed to subtract days from date key")?
    };

    Ok(shifted.format("%Y-%m-%d").to_string())
}

pub fn last_n_date_keys(now: DateTime<Utc>, time_zone: &str, count: usize) -> Result<Vec<String>> {
    let time_zone: Tz = time_zone
        .parse()
        .with_context(|| format!("unsupported time zone: {time_zone}"))?;
    let today = now.with_timezone(&time_zone).date_naive();

    Ok((0..count)
        .map(|index| {
            today
                .checked_sub_days(Days::new(index as u64))
                .unwrap_or(today)
                .format("%Y-%m-%d")
                .to_string()
        })
        .collect())
}

pub fn month_key_for(now: DateTime<Utc>, time_zone: &str) -> Result<String> {
    let time_zone: Tz = time_zone
        .parse()
        .with_context(|| format!("unsupported time zone: {time_zone}"))?;
    Ok(now.with_timezone(&time_zone).format("%Y-%m").to_string())
}

File: electron/services/dateKeys.ts
export function parseTimestamp(value: string): Date | null {
  const date = new Date(value);
  return isNaN(date.getTime()) ? null : date;
}

export function dateKeyFor(date: Date, timeZone: string): string {
  try {
    const formatter = new Intl.DateTimeFormat("en-CA", {
      timeZone,
      year: "numeric",
      month: "2-digit",
      day: "2-digit"
    });
    return formatter.format(date);
  } catch (error) {
    throw new Error(`unsupported time zone: ${timeZone}`);
  }
}

export function monthKeyFor(date: Date, timeZone: string): string {
  const dateKey = dateKeyFor(date, timeZone);
  return dateKey.slice(0, 7);
}

export function addDaysToDateKey(dateKey: string, deltaDays: number): string {
  const parts = dateKey.split("-").map(Number);
  if (parts.length !== 3 || parts.some(isNaN)) {
    throw new Error(`unsupported date key: ${dateKey}`);
  }
  const [year, month, day] = parts;
  const date = new Date(Date.UTC(year, month - 1, day));
  date.setUTCDate(date.getUTCDate() + deltaDays);

  const y = date.getUTCFullYear();
  const m = String(date.getUTCMonth() + 1).padStart(2, "0");
  const d = String(date.getUTCDate()).padStart(2, "0");
  return `${y}-${m}-${d}`;
}

export function lastNDateKeys(now: Date, timeZone: string, count: number): string[] {
  const todayKey = dateKeyFor(now, timeZone);
  const result: string[] = [];
  for (let i = 0; i < count; i++) {
    result.push(addDaysToDateKey(todayKey, -i));
  }
  return result;
}

export function formatUtcTimestamp(date: Date): string {
  return date.toISOString();
}
